File: C/integration.c
#include "integration.h"

double left_rule(double (*f)(double), double a, double b, int n) {
	int i;
	double sum = 0;
	double h = (b - a) / n;
	
	for (i = 0; i < n; ++i) {
		sum += f(a + i * h);
	}
	return h * sum;
}

double right_rule(double (*f)(double), double a, double b, int n) {
	int i;
	double sum = 0;
	double h = (b - a) / n;
	
	for (i = 1; i <= n; ++i) {
		sum += f(a + i * h);
	}
	return h * sum;
}

double midpoint_rule(double (*f)(double), double a, double b, int n) {
	double i;
	double sum = 0;
	double h = (b - a) / n;
	
	for (i = 0.5; i < n; ++i) {
		sum += f(a + i * h);
	}
	return h * sum;
}

double trapezoidal_rule(double (*f)(double), double a, double b, int n) {
	int i;
	double sum = 0;
	double h = (b - a) / n;
	
	sum += f(a);
	for (i = 1; i < n; ++i) {
		sum += 2 * f(a + i * h);
	}
	sum += f(b);
	
	return h / 2 * sum;
}

double simpson_rule(double (*f)(double), double a, double b, int n) {
	int i;
	double sum = 0;
	double h = (b - a) / n;
	
	sum += f(a);
	for (i = 1; i < n; ++i) {
		if (i % 2 == 0) {
			sum += 2 * f(a + i * h);
		}
		else {
			sum += 4 * f(a + i * h);
		}	
	}
	sum += f(b);
	
	return h / 3 * sum;
}

File: integration.py
def left_rule(f, a, b, n):
	h = (b - a) / n
	area = 0
	for i in range(n):
		area += f(a + i * h)
	return h * area

def right_rule(f, a, b, n):
	h = (b - a) / n
	area = 0
	for i in range(1, n + 1):
		area += f(a + i * h)
	return h * area

def midpoint_rule(f, a, b, n):
	h = (b - a) / n
	area = 0
	for i in range(n):
		area += f(a + (i + 0.5) * h)
	return h * area

def trapezoidal_rule(f, a, b, n):
	h = (b - a) / n
	area = 0
	
	area += f(a)
	for i in range(1, n):
		area += 2 * f(a + i * h)
	area += f(b)
	return h / 2 * area

def simpsons_rule(f, a, b, n):
	h = (b - a) / n
	area = 0
	
	area += f(a)
	for i in range(1, n):
		if i % 2 == 1:
			area += 4 * f(a + i * h)
		else:
			area += 2 * f(a + i * h)
	area += f(b)
	return h / 3 * area
	